// @chakra-ui/cli@3.36.0/dist/esm/utils/io.js
"use strict";
import { log } from '@clack/prompts';
import chokidar from 'chokidar';
import createDebug from 'debug';
import { build } from 'esbuild';
import { existsSync, mkdirSync, rm, realpathSync } from 'node:fs';
import { writeFile } from 'node:fs/promises';
import { createRequire } from 'node:module';
import { dirname, resolve, extname, join } from 'node:path';
import vm from 'node:vm';
import { resolveTsconfig } from './resolve-tsconfig.js';

const debug = createDebug("chakra:io");
const require$1 = createRequire(import.meta.url);
async function bundleFile(file, cwd, tsconfigPath) {
  const tsconfig = await resolveTsconfig(file, tsconfigPath);
  debug("resolved tsconfig for esbuild:", tsconfig);
  const result = await build({
    platform: "node",
    format: "cjs",
    mainFields: ["module", "main"],
    absWorkingDir: cwd,
    entryPoints: [file],
    outfile: "out.js",
    write: false,
    bundle: true,
    sourcemap: false,
    metafile: true,
    ...tsconfig ? { tsconfig } : {}
  });
  const { text } = result.outputFiles[0];
  return {
    code: text,
    dependencies: result.metafile ? Object.keys(result.metafile.inputs) : []
  };
}
function loadBundledCode(file, code) {
  try {
    return loadViaRequire(file, code);
  } catch {
    return loadViaVm(code);
  }
}
function loadViaRequire(file, code) {
  const ext = extname(file);
  const realFileName = realpathSync.native(file);
  const defaultLoader = require$1.extensions[ext];
  require$1.extensions[ext] = (mod, filename) => {
    if (filename === realFileName) {
      mod._compile(code, filename);
    } else {
      defaultLoader?.(mod, filename);
    }
  };
  try {
    delete require$1.cache[require$1.resolve(file)];
    return require$1(file);
  } finally {
    require$1.extensions[ext] = defaultLoader;
  }
}
function loadViaVm(code) {
  const mod = { exports: {} };
  const ctx = vm.createContext({
    module: mod,
    exports: mod.exports,
    require: require$1
  });
  vm.runInContext(code, ctx);
  return mod.exports;
}
const systemExports = ["default", "preset", "system"];
const isRecord = (mod) => {
  return typeof mod === "object" && mod !== null;
};
const isValidSystem = (mod) => {
  return isRecord(mod) && Object.hasOwnProperty.call(mod, "$$chakra");
};
function resolveSystemExport(mod) {
  const candidate = isRecord(mod) && mod.default != null ? mod.default : mod;
  if (!isRecord(candidate) || isValidSystem(candidate)) return candidate;
  const exports$1 = candidate;
  return exports$1.default || exports$1.preset || exports$1.system || candidate;
}
function formatExportName(name) {
  return `"${name}"`;
}
function getExportNames(mod) {
  if (!isRecord(mod) || isValidSystem(mod)) return [];
  return Object.keys(mod).filter((name) => name !== "__esModule");
}
function formatInvalidSystemError(file, mod) {
  const exportNames = getExportNames(mod);
  const foundExports = exportNames.length ? `Found export${exportNames.length === 1 ? "" : "s"}: ${exportNames.map(formatExportName).join(", ")}.` : "Found no named exports.";
  return [
    `No Chakra system export found in ${file}.`,
    `The chakra typegen command expects ${systemExports.map(formatExportName).join(
      ", "
    )}, or a CommonJS export to be a Chakra system created with createSystem(...).`,
    foundExports,
    "If this file exports a config from defineConfig(...), wrap it first: export default createSystem(defaultConfig, config)."
  ].join(" ");
}
const read = async (file, options = {}) => {
  const { cwd = process.cwd(), tsconfig } = options;
  const filePath = resolve(file);
  const bundle = await bundleFile(filePath, cwd, tsconfig);
  const mod = loadBundledCode(filePath, bundle.code);
  const resolvedMod = resolveSystemExport(mod);
  if (!isValidSystem(resolvedMod)) {
    throw new Error(formatInvalidSystemError(file, mod));
  }
  return { mod: resolvedMod, dependencies: bundle.dependencies };
};
const outPath = (path, file) => {
  const ext = process.env.LOCAL ? "ts" : "d.ts";
  return join(path, `${file}.${ext}`);
};
function ensureDir(dirPath) {
  if (existsSync(dirPath)) return;
  ensureDir(dirname(dirPath));
  mkdirSync(dirPath);
}
const write = async (path, file, content) => {
  try {
    await writeFile(outPath(path, file), await content);
  } catch (error) {
    throw new Error(
      `Failed to write file ${outPath(path, file)}: ${error instanceof Error ? error.message : String(error)}`
    );
  }
};
function watch(paths, cb) {
  const watcher = chokidar.watch(paths, { ignoreInitial: true });
  watcher.on("ready", cb).on("change", async (filePath) => {
    log.info(`\u{1F4E6} File changed: ${filePath}`);
    return cb();
  });
  process.once("SIGINT", () => watcher.close());
  process.once("SIGTERM", () => watcher.close());
}
async function clean(basePath) {
  log.info("\u{1F9F9} Cleaning output directory");
  rm(basePath, { recursive: true }, (err) => {
    if (err) {
      log.error(err.message);
    }
  });
}

export { clean, ensureDir, read, watch, write };
